twilight-gateway-queue/src/in_memory.rs

```diff
@@ -0,0 +1,219 @@
+    reset_at: Instant,
+    /// The number of permits to reset to.
+    total: u16,
+}
+
+/// [`InMemoryQueue`]'s background task runner.
+///
+/// Buckets requests such that only one timer is necessary.
+async fn runner(
+    mut rx: mpsc::UnboundedReceiver<Message>,
+    Settings {
+        mut max_concurrency,
+        mut remaining,
+        reset_at,
+        mut total,
+    }: Settings,
+) {
+    let reset_at = sleep_until(reset_at);
+    let interval = sleep_until(Instant::now());
+    tokio::pin! {
+        reset_at,
+        interval
+    };
+    let create_queues = |max_concurrency: u8| {
+        iter::repeat_with(VecDeque::new)
+            .take(max_concurrency.into())
+            .collect::<Vec<_>>()
+    };
+    let mut queues = create_queues(max_concurrency);
+
+    'outer: loop {
+        tokio::select! {
+            biased;
+            _ = &mut reset_at => {
+                remaining = total;
+                reset_at.as_mut().reset(Instant::now() + Duration::from_secs(60 * 60 * 24));
+            }
+            message = rx.recv() => {
+                match message {
+                    Some(Message::Request{shard, tx}) => {
+                        if max_concurrency == 0 {
+                            _ = tx.send(());
+                        } else {
+                            #[allow(clippy::cast_possible_truncation)]
+                            queues[(shard % u64::from(max_concurrency)) as usize].push_back((shard, tx));
+                        }
+                    }
+                    Some(Message::Update(update)) => {
+                        let deadline;
+                        Settings {max_concurrency, remaining, reset_at: deadline, total} = update;
+
+                        if queues.len() != max_concurrency as usize {
+                            let unbalanced = queues.into_iter().flatten();
+                            queues = create_queues(max_concurrency);
+                            for (shard, tx) in unbalanced {
+                                #[allow(clippy::cast_possible_truncation)]
+                                queues[(shard % u64::from(max_concurrency)) as usize].push_back((shard, tx));
+                            }
+                        }
+                        reset_at.as_mut().reset(deadline);
+                    }
+                    None => break,
+                }
+            }
+            _ = &mut interval, if queues.iter().any(|queue| !queue.is_empty()) => {
+                let now = Instant::now();
+                let span = tracing::info_span!("bucket", capacity = %queues.len(), ?now);
+                interval.as_mut().reset(now + IDENTIFY_INTERVAL);
+                for (rate_limit_key, queue) in queues.iter_mut().enumerate() {
+                    if remaining == 0 {
+                        (&mut reset_at).await;
+                        remaining = total;
+                        reset_at.as_mut().reset(now + Duration::from_secs(60 * 60 * 24));
```

Comment:
same here, move this to a constant and avoid the use of `now`. This logic seems flawed too, since `now` is before `reset_at` has been reached and therefore the new `reset_at` calculated might be in the past unless I'm mistaken